provider/pkg/provider/provider.go

```diff
@@ -1035,46 +1037,73 @@ func (p *cfnProvider) Update(ctx context.Context, req *pulumirpc.UpdateRequest)
 		return nil, errors.Wrapf(err, "failed to parse inputs for update")
 	}
 
-	diff, err := p.diffState(req.GetOlds(), newInputs, label)
+	oldState, err := plugin.UnmarshalProperties(req.GetOlds(), plugin.MarshalOptions{
+		Label:        fmt.Sprintf("%s.oldState", label),
+		KeepUnknowns: true,
+		RejectAssets: true,
+		KeepSecrets:  true,
+	})
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "diff failed because malformed resource inputs")
 	}
+	oldInputs := parseCheckpointObject(oldState)
```

Comment:
Looks like this is new behavior on existing resources not just the Extension Resource; curious about the motivation for adding it.

Reply:
Ah yes, this is currently unused on the schema resource condition. This is because the `p.diffState()` method takes the grpc type, the calls parseCheckpointObject itself.

I avoided initially touching the `diffState` method, but think this warants a clean-up to avoid doing the work twice. This means that we can completely remove `diffState` as it's literally just doing the unpacking of the `oldInputs` then calling diff.